imap_processing/swe/l1a_swe.py

```diff
@@ -0,0 +1,183 @@
+        return 0
+
+def get_step_indexes(spin_number):
+    # To get one full data, we need to get data from four spins where each spin
+    # data is stored in one packet data. ESA_STEPS from metadata gives information about
+    # which spin data is stored in which packet.
+    # ESA_STEPS = 0 --> first spin data
+    # ESA_STEPS = 1 --> second spin data
+    # ESA_STEPS = 2 --> third spin data
+    # ESA_STEPS = 3 --> fourth spin data
+
+    # These indexes is where each spin's data goes in full data table.
+    # It's like putting a puzzle together. One full data table in the
+    # the algorithm document is of this shape ( 24, 30, 7). Once we
+    # have populated data table with all four spins, we can reshape
+    # it to what subsequent algorithm needs which is (24, 7, 30).
+    spin_one_indexes = [[1,  5, 9, 13, 17,  21], [23, 19, 15, 11, 7, 3]]
+    spin_two_indexes = [[2,  6, 10, 14, 18,  22], [24, 20, 16, 12, 8, 4]]
+    spin_three_indexes = [[3,  7, 11, 15, 19,  23], [21, 17, 13, 9, 5, 1]]
+    spin_four_indexes = [[4,  8, 12, 16, 20,  24], [22, 18, 14, 10, 6, 2]]
+
+    match spin_number:
+        case 0:
+            return spin_one_indexes
+        case 1:
+            return spin_two_indexes
+        case 2:
+            return spin_three_indexes
+        case 3:
+            return spin_four_indexes
+        case _:
+            print(f"Error! Invalid ESA_STEPS - {spin_number}")
+
+
+def swe_l1a():
+    """SWE L1A algorithm steps:
+        - Read data from SWE packet file
+        - Uncompress data
+        - Store metadata and data in attrs and DataArray of xarray respectively
+        - Save complete data to cdf file
+    Each L1A data will have this shape: 24 rows, 7 columns, and each cell in 24 x 7 table contains
+    30 element array. These dimension maps to this:
+        24 rows --> 24 energy steps
+        7 column --> 7 CEMs value
+        30 element --> 30 spin angles
+    """
+    packet_file = f"{packet_definition_directory}/../swe/tests/science_block_20221116_163611Z_idle.bin"
+    decom_data = decom_swe.decom_packets(packet_file)
+
+    total_packets = 0
+    one_full_cycle = np.zeros((24, 30, 7))
+    while total_packets < len(decom_data):
+        # If ESA_STEPS is 0, we are getting data for new cycle
+        spin_number = decom_data[total_packets].data["ESA_STEPS"].raw_value
+        if spin_number == 0:
+            # It should follow by 4 packets where each packet contains
+            # one spin's data.
+            one_full_cycle = np.zeros((24, 30, 7))
+
+        # TODO: get metadata of each data and stores. Find out how to combine four
+        # packet's metadata.
+
+        # read raw data
+        binary_data = decom_data[total_packets].data["SCIENCE_DATA"].raw_value
+        # read raw data as binary array using bitstring
+        bit_array = bitstring.ConstBitStream(bin=binary_data)
+        # chunk binary into 1260 units each with 8-bits
+        byte_data = bit_array.readlist(["uint:8"] * 1260)
```

Comment:
Alternatively if you wanted to go straight to a numpy array and avoid bitstrings since you are dealing with full bytes.
```suggestion
        binary_data = decom_data[total_packets].data["SCIENCE_DATA"].raw_value
        # read raw data into a uint8 numpy array
        byte_data = np.frombuffer(binary_data, dtype=np.uint8)
```

Reply:
good to know!

Reply:
So, I decided to run some timing comparisons to see if this would be worth it actually...

```bash
python -m timeit 'import bitstring; bitstring.ConstBitStream(bytes(1260)).readlist(["uint:8"]*1260)'
100 loops, best of 5: 2.68 msec per loop
python -m timeit 'import numpy as np; np.frombuffer(bytes(1260), dtype=np.uint8)'
500000 loops, best of 5: 668 nsec per loop
```

That is a pretty massive improvement by using numpy directly. I'm kind of surprised bitstring is that slow...

Reply:
That's impressive!